module_utils/foreman_helper.py

```diff
@@ -51,22 +53,199 @@ def filter_module_params(self):
+        entity_id = old_entity.pop('id')
+        volatile_entity = old_entity.copy()
+        result = volatile_entity
+        fields = []
+        for key, value in volatile_entity.items():
+            if key in entity_dict:
+                new_value = entity_dict[key]
+                if check_type and not isinstance(value, type(new_value)):
+                    if isinstance(value, bool):
+                        new_value = boolean(new_value)
+                    elif isinstance(value, list):
+                        new_value = str(sorted(new_value.split(',')))
+                        value = str(sorted([str(v) for v in value]))
+                    else:
+                        if value is None:
+                            value = ""
+                        new_value = type(value)(new_value)
+                if isinstance(value, list) and value and isinstance(value[0], dict) and 'id' in value[0]:
+                    value = [{'id': item['id']} for item in value]
+                elif isinstance(value, dict) and 'id' in value:
+                    value = value['id']
+                if isinstance(new_value, dict) and 'id' in new_value:
+                    new_value = new_value['id']
+                if not value == new_value:
+                    volatile_entity[key] = new_value
+                    fields.append(key)
+            # check_missing is a special case, Foreman sometimes returns different values
+            # depending on what 'type' of same object you are requesting. Content View
+            # Filters are a prime example. We list these attributes in `check_missing`
+            # so we can ensure the entity is as the user specified.
+            if check_missing is not None and key not in entity_dict and key in check_missing:
+                volatile_entity[key] = None
+                fields.append(key)
+        if force_update:
+            for key in force_update:
+                fields.append(key)
+        if check_missing is not None:
+            for key in check_missing:
+                if key in entity_dict and key not in volatile_entity.keys():
+                    volatile_entity[key] = new_value
+                    fields.append(key)
+        if len(fields) > 0:
+            new_data = {}
+            for key, value in volatile_entity.items():
+                if key in fields:
+                    new_data[key] = value
+            return self._resource_action(resource, 'update', self._generate_resource_params(resource, 'update', resource_id=entity_id, params=new_data))
+        return False, result
+
+    def ensure_resource_state(self, resource, entity_dict, entity, state, check_missing=None, check_type=None, force_update=None):
+        changed, _ = self.ensure_resource(resource, entity_dict, entity, state, check_missing, check_type, force_update)
+        return changed
+
+    def ensure_resource(self, resource, entity_dict, entity, state, check_missing=None, check_type=None, force_update=None):
```

Comment:
I can only speculate, you didn't like the name `naildown_entity` enough...

Reply:
I just didn't want the problem to look like a nail ;)